Add disabling GRC known issue (#5180)
open-cluster-management/rhacm-docs: release_notes/known_issues_governance.adoc

- Another controller or process is also updating the object with different values.
+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the values are different, another controller or process might be updating them. Check the `metadata` of the object to help identify why the values are different.

- The `objectDefinition` in the `ConfigurationPolicy` does not match because of Kubernetes processing the object when the policy is applied.
+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the keys are different or missing, Kubernetes might have processed the keys before applying them to the object, such as removing keys containing default or empty values.
+
*Note:* If `pruneObjectBehavior` is set to something other than `none`, disabling the policy causes the objects to be cleaned up. In this case, set `prunObjectBehavior` to `none` so that the objects exist after the policy is disabled.

> For example, the `stringData` map in a `Secret` resource is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.

Reviewer: ```suggestion

For example, the `stringData` map in a `Secret` resource is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.
```

[#policy-status-repeated-updates]
== Policy status shows repeated updates when enforced

If a policy is set to `remediationAction: enforce` and is repeatedly updated, the {product-title-short} console shows repeated violations with successful updates. This might happen in the following two cases:

> - Another controller or process is also updating the object with different values.

Reviewer 2: I expected a list of "cases" but was interrupted with unexpected resolutions. How about introducing them differently?

See the following two possible causes and solutions for the error: 

Or something like that ^^^ to tell the user both are following (imagine a screen reader).

Reviewer: updated

+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the values are different, another controller or process might be updating them. Check the `metadata` of the object to help identify why the values are different.

- The `objectDefinition` in the `ConfigurationPolicy` does not match because of Kubernetes processing the object when the policy is applied.
+

> To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the keys are different or missing, Kubernetes might have processed the keys before applying them to the object, such as removing keys containing default or empty values.

Reviewer 2: a fair use of past participle that we may want to show Faisa, because once in a while we have to use it.

+

> *Note:* If `pruneObjectBehavior` is set to something other than `none`, disabling the policy causes the objects to be cleaned up. In this case, set `prunObjectBehavior` to `none` so that the objects exist after the policy is disabled.

Reviewer 3: ```suggestion
*Note:* If `pruneObjectBehavior` is set to something other than `None`, disabling the policy causes the objects to be cleaned up. In this case, set `pruneObjectBehavior` to `None` so that the objects exist after the policy is disabled.
```

Reviewer: Great catch!

For example, the `stringData` map in a `Secret` resource is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.

[#psp-not-supported-ocp]

> == Pod security policies not supported on OpenShift 4.12 and later

Reviewer 2: I think we need {ocp-short} here for proper product reference.

> == Pod security policies not supported on OpenShift 4.12 and later

Reviewer: ```suggestion
== Pod security policies not supported on {ocp-short} 4.12 and later
```

[#psp-not-supported-ocp]
== Pod security policies not supported on {ocp-short} 4.12 and later

The support of pod security policies is removed from {ocp-short} 4.12 and later, and from Kubernetes v1.25 and later. If you apply a `PodSecurityPolicy` resource, you might receive the following non-compliant message:

----
violation - couldn't find mapping resource with kind PodSecurityPolicy, please check if you have CRD deployed
----

[#disabling-grc]
== Governance deployments do not shut down gracefully when disabled

When you disable governance deployments in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. Complete the following steps to disable governance so that the deployments also get cleaned up:

. Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If you do this for all managed clusters, run the following command:
+
[source,bash]
----
for CLUSTER in $(oc get managedclusters -o jsonpath='{.items[].metadata.name}'); do
  oc patch -n ${CLUSTER} klusterletaddonconfig ${CLUSTER} --type=merge --patch='{"spec":{"policyController":{"enabled":false}}}'
done
----

> . (For self-managed hub clusters only) Delete the `ManifestWork` for the self-managed hub cluster and remove the finalizer on the `ManagedClusterAddon` if the `governance-policy-framework-uninstall` pod of a self-managed hub cluster is in `CrashLoopBackOff`. Run the following commands:

Reviewer: ```suggestion
. For self-managed hub clusters only: Delete the `ManifestWork` for the self-managed hub cluster and remove the finalizer on the `ManagedClusterAddon` if the `governance-policy-framework-uninstall` pod of a self-managed hub cluster is in `CrashLoopBackOff`. Run the following commands:
```

+
[source,bash]
----
oc delete manifestwork -n local-cluster -l open-cluster-management.io/addon-name=governance-policy-framework
oc patch managedclusteraddon -n local-cluster governance-policy-framework --type=merge --patch='{"metadata":{"finalizers":[]}}'
----

> . Now that governance is disabled on the managed clusters, you can disable globally if required by setting the `grc` element under `spec.overrides` to `false` in the `MultiClusterHub` object. Run the following command:

Reviewer: ```suggestion
. Disable governance globally, if required, by setting the `grc` element in the `spec.overrides` section to `false` in the `MultiClusterHub` object. Run the following command:
```

+
[source,bash]
----
oc edit multiclusterhub <name> -n <namespace>
----

> . (For self-managed hubs only) If there are any local cluster policies, you can delete the policies by running the following command:

Reviewer: ```suggestion
. For self-managed hubs only: If there are any local cluster policies, you can delete the policies by running the following command:
```

+
[source,bash]
----
oc delete policies -n local-cluster --all
----

> . To reenable governance in the `KlusterletAddonConfig`, reenable the `grc` element of the `spec.overrides` in the `MultiClusterHub` first and then run the loop with `"enabled":true`. If the deployments are unsuccessful, the `governance-policy-addon-controller` might have a stale lease. Delete the lease by using the following command:

Reviewer: ```suggestion
. To re-enable governance in the `KlusterletAddonConfig`, re-enable the `grc` element of the `spec.overrides` section in the `MultiClusterHub`. Run the following command:
+
[source,bash]
----
for CLUSTER in $(oc get managedclusters -o jsonpath='{.items[].metadata.name}'); do
  oc patch -n ${CLUSTER} klusterletaddonconfig ${CLUSTER} --type=merge --patch='{"spec":{"policyController":{"enabled":true}}}'
done
----

. If the deployments are unsuccessful, the `governance-policy-addon-controller` might have a stale lease. Delete the lease by using the following command:
```

[#policy-status-repeated-updates]
== Policy status shows repeated updates when enforced

If a policy is set to `remediationAction: enforce` and is repeatedly updated, the {product-title-short} console shows repeated violations with successful updates. This might happen in the following two cases:

- Another controller or process is also updating the object with different values.
+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the values are different, another controller or process might be updating them. Check the `metadata` of the object to help identify why the values are different.

> - The `objectDefinition` in the `ConfigurationPolicy` does not match because Kubernetes processing the object when the policy is applied.

Reviewer 4: ```suggestion
- The `objectDefinition` in the `ConfigurationPolicy` does not match because of Kubernetes processing the object when the policy is applied.
```

+

> To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the keys are different or missing, Kubernetes might have processed the keys before applying them to the object, such as removing keys containing default or empty values.

Reviewer 4: After thinking about this, disabling the policy might be dangerous because of `pruneObjectBehavior`. Perhaps add a warning here or you could recommend increasing the evaluation interval.

Author: That's a good point. I'll add a note.

+
Known examples:
+
|===
| Kind | Issue description

|`Secret`

> | The `stringData` map is processed by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.

Reviewer 4: ```suggestion
| The `stringData` map is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.
```

|===

[#psp-not-supported-ocp]
== Pod security policies not supported on OpenShift 4.12 and later

The support of pod security policies is removed from {ocp-short} 4.12 and later, and from Kubernetes v1.25 and later. If you apply a `PodSecurityPolicy` resource, you might receive the following non-compliant message:

----
violation - couldn't find mapping resource with kind PodSecurityPolicy, please check if you have CRD deployed
----

[#disabling-grc]
== Governance deployments do not shut down gracefully when disabled

> When disabling the Governance deployments, if Governance is disabled in the MultiClusterHub object, the deployments do not all get cleaned up properly. To disable Governance so that the deployments also get cleaned up:

Reviewer 4: I'm not sure if this is required.
```suggestion
When disabling the Governance deployments, if Governance is disabled in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. To disable Governance so that the deployments also get cleaned up:
```

Author: Are you wondering whether disabling the `KlusterletAddonConfig` is required?

Reviewer 4: Sorry, I meant the back ticks

Author: Ah, gotcha. I like them--I think it's good for consistency and to signal that it's a code reference.

> . Disable the `policyController` in the KlusterletAddonConfig for the managed cluster. If doing this for all managed clusters, this loop will handle it:

Reviewer 4: ```suggestion
. Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If doing this for all managed clusters, this loop will handle it:
```

[#policy-status-repeated-updates]
== Policy status shows repeated updates when enforced

If a policy is set to `remediationAction: enforce` and is repeatedly updated, the {product-title-short} console shows repeated violations with successful updates. This might happen in the following two cases:

- Another controller or process is also updating the object with different values.
+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the values are different, another controller or process might be updating them. Check the `metadata` of the object to help identify why the values are different.

- The `objectDefinition` in the `ConfigurationPolicy` does not match because of Kubernetes processing the object when the policy is applied.
+
To resolve the issue, disable the policy and compare the differences between `objectDefinition` in the policy and the object on the managed cluster. If the keys are different or missing, Kubernetes might have processed the keys before applying them to the object, such as removing keys containing default or empty values.
+

> *NOTE:* If `pruneObjectBehavior` has been set to something other than `none`, disabling the policy will cause the objects to be cleaned up. In this case, set `prunObjectBehavior` to `none` so that the objects exist after the policy is disabled.

Reviewer: ```suggestion
*Note:* If `pruneObjectBehavior` is set to something other than `none`, disabling the policy causes the objects to be cleaned up. In this case, set `prunObjectBehavior` to `none` so that the objects exist after the policy is disabled.
```

> +
> Known examples:
> +
> |===
> | Kind | Issue description
> 
> |`Secret`
> | The `stringData` map is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.
> 
> |===

Reviewer: I really like the idea of using the table, but because there is only one example we should not use a table. In the case that we decide to list another example, let's go with the table approach.

```suggestion
For example, the `stringData` map in a `Secret` resource is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.
```

[#psp-not-supported-ocp]
== Pod security policies not supported on OpenShift 4.12 and later

The support of pod security policies is removed from {ocp-short} 4.12 and later, and from Kubernetes v1.25 and later. If you apply a `PodSecurityPolicy` resource, you might receive the following non-compliant message:

----
violation - couldn't find mapping resource with kind PodSecurityPolicy, please check if you have CRD deployed
----

[#disabling-grc]
== Governance deployments do not shut down gracefully when disabled

> When disabling the Governance deployments, if Governance is disabled in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. To disable Governance so that the deployments also get cleaned up:

Reviewer: ```suggestion
When you disable governance deployments in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. Complete the following steps to disable governance so that the deployments also get cleaned up:
```

> . Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If doing this for all managed clusters, this loop will handle it:

Reviewer: ```suggestion
. Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If you do this for all managed clusters, run the following command:
```

+
[source,bash]
----
for CLUSTER in $(oc get managedclusters -o jsonpath='{.items[].metadata.name}'); do
  oc patch -n ${CLUSTER} klusterletaddonconfig ${CLUSTER} --type=merge --patch='{"spec":{"policyController":{"enabled":false}}}'
done
----

> . (For self-managed hubs only) A self-managed hub's `governance-policy-framework-uninstall` pod will go into `CrashLoopBackOff` because it is in conflict with the `governance-policy-propagator` pod. If this is the case, delete the `ManifestWork` for the self-managed hub and remove the finalizer on the `ManagedClusterAddon`:

Reviewer: ```suggestion
. (For self-managed hub clusters only) Delete the `ManifestWork` for the self-managed hub cluster and remove the finalizer on the `ManagedClusterAddon` if the `governance-policy-framework-uninstall` pod of a self-managed hub cluster is in `CrashLoopBackOff`. Run the following commands:
```

+
[source,bash]
----
oc delete manifestwork -n local-cluster -l open-cluster-management.io/addon-name=governance-policy-framework
oc patch managedclusteraddon -n local-cluster governance-policy-framework --type=merge --patch='{"metadata":{"finalizers":[]}}'
----

> . Now that Governance is disabled on the managed clusters, it can be disabled globally if required by setting the `grc` element under `spec.overrides` to `false` in the `MultiClusterHub` object:

Reviewer: ```suggestion
. Now that governance is disabled on the managed clusters, you can disable globally if required by setting the `grc` element under `spec.overrides` to `false` in the `MultiClusterHub` object. Run the following command:
```

+
[source,bash]
----
oc edit multiclusterhub <name> -n <namespace>
----

> . (For self-managed hubs only) If there were any local-cluster policies, those can optionally be deleted:

Reviewer: ```suggestion
. (For self-managed hubs only) If there are any local cluster policies, you can delete the policies by running the following command:
```

+
[source,bash]
----
oc delete policies -n local-cluster --all
----

> To re-enable Governance after doing this, re-enable the `grc` element under `spec.overrides` in the `MultiClusterHub` first and then run the loop with `"enabled":true` to re-enable Governance in the KlusterletAddonConfig. If the deployments don't come up successfully, the `governance-policy-addon-controller` might have a stale lease. Deleting the lease will allow it to function properly again:

Reviewer: ```suggestion
. To reenable governance in the `KlusterletAddonConfig`, reenable the `grc` element of the `spec.overrides` in the `MultiClusterHub` first and then run the loop with `"enabled":true`. If the deployments are unsuccessful, the `governance-policy-addon-controller` might have a stale lease. Delete the lease by using the following command:
```

[#policy-status-repeated-updates]
== Policy status shows repeated updates when enforced

> If a policy is set to `remediationAction: enforce` and is repeatedly updated, the {product-title-short} console shows repeated violations with successful updates. This might happen in the following two cases:

Reviewer: ```suggestion
If a policy is set to `remediationAction: enforce` and is repeatedly updated, the {product-title-short} console shows repeated violations with successful updates. See the following two possible causes and solutions for the error:
```

+
*Note:* If `pruneObjectBehavior` is set to something other than `None`, disabling the policy causes the objects to be cleaned up. In this case, set `pruneObjectBehavior` to `None` so that the objects exist after the policy is disabled.

For example, the `stringData` map in a `Secret` resource is converted by Kubernetes to `data` with `base64` encoded values. Instead of using `stringData`, use `data` directly with `base64` encoded values instead of strings.

[#psp-not-supported-ocp]
== Pod security policies not supported on {ocp-short} 4.12 and later

The support of pod security policies is removed from {ocp-short} 4.12 and later, and from Kubernetes v1.25 and later. If you apply a `PodSecurityPolicy` resource, you might receive the following non-compliant message:

----
violation - couldn't find mapping resource with kind PodSecurityPolicy, please check if you have CRD deployed
----

[#disabling-grc]

> == Governance deployments do not shut down gracefully when disabled

Reviewer: ```suggestion
== Governance deployments do not shut down without error when disabled
```

> == Governance deployments do not shut down without error when disabled

Reviewer: ```suggestion
== Governance deployments do not shut down without errors when disabled
```

> When you disable governance deployments in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. Complete the following steps to disable governance so that the deployments also get cleaned up:

Reviewer: ```suggestion
When you disable governance deployments in the `MultiClusterHub` object, the deployments are not cleaned without errors. Complete the following steps to disable governance so that the deployments also get cleaned up:
```

. Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If you do this for all managed clusters, run the following command:
+
[source,bash]
----
for CLUSTER in $(oc get managedclusters -o jsonpath='{.items[].metadata.name}'); do
  oc patch -n ${CLUSTER} klusterletaddonconfig ${CLUSTER} --type=merge --patch='{"spec":{"policyController":{"enabled":false}}}'
done
----

> . For self-managed hub clusters only: Delete the `ManifestWork` for the self-managed hub cluster and remove the finalizer on the `ManagedClusterAddon` if the `governance-policy-framework-uninstall` pod of a self-managed hub cluster is in `CrashLoopBackOff`. Run the following commands:

Reviewer: From @swopebe is self-managed hub cluster the local cluster? @JustinKuli

Reviewer 3: Yes. When the hub is self-managing, its name is always "local-cluster".

----
violation - couldn't find mapping resource with kind PodSecurityPolicy, please check if you have CRD deployed
----

[#disabling-grc]
== Governance deployments do not shut down gracefully when disabled

> When you disable governance deployments in the `MultiClusterHub` object, the deployments do not all get cleaned up properly. Complete the following steps to disable governance so that the deployments also get cleaned up:

Reviewer 2: I think we need another word here, other than gracefully. Like "deployments do not shut down completely" or "do not shut down without errors"....

Gracefully makes it seem like the deployments are dancing, but not gracefully.

Reviewer 2: "Get cleaned up" is also a bit jarring.

What are we trying to say here? That they are not shut down completely?

Reviewer 2: Or that they do not shut down correctly?

Reviewer 2: properly?

Reviewer: @swopebe from my perspective, "do not shut down without errors" is most accurate @JustinKuli thoughts?

Reviewer 2: I think that is fair and should be technically correct.

. Disable the `policyController` in the `KlusterletAddonConfig` for the managed cluster. If you do this for all managed clusters, run the following command:
+
[source,bash]
----
for CLUSTER in $(oc get managedclusters -o jsonpath='{.items[].metadata.name}'); do
  oc patch -n ${CLUSTER} klusterletaddonconfig ${CLUSTER} --type=merge --patch='{"spec":{"policyController":{"enabled":false}}}'
done
----

> . (For self-managed hub clusters only) Delete the `ManifestWork` for the self-managed hub cluster and remove the finalizer on the `ManagedClusterAddon` if the `governance-policy-framework-uninstall` pod of a self-managed hub cluster is in `CrashLoopBackOff`. Run the following commands:

Reviewer 2: This can be a line with a colon:

For self-managed hub clusters only: (or a comma instead of colon. I think we normally use the colon)

Reviewer 2: Isn't "self-managed hub cluster" called the local cluster?

+
[source,bash]
----
oc delete manifestwork -n local-cluster -l open-cluster-management.io/addon-name=governance-policy-framework
oc patch managedclusteraddon -n local-cluster governance-policy-framework --type=merge --patch='{"metadata":{"finalizers":[]}}'
----

> . Now that governance is disabled on the managed clusters, you can disable globally if required by setting the `grc` element under `spec.overrides` to `false` in the `MultiClusterHub` object. Run the following command:

Reviewer 2: We don't need "Now that" or "next" or anything....that is all extra. We just give them the next step; that is the most imperative.

+
[source,bash]
----
oc edit multiclusterhub <name> -n <namespace>
----

> . (For self-managed hubs only) If there are any local cluster policies, you can delete the policies by running the following command:

Reviewer 2: Same here with

For self-managed only: (Though I think we mean For the local cluster only?)

+
[source,bash]
----
oc delete policies -n local-cluster --all
----

> . To reenable governance in the `KlusterletAddonConfig`, reenable the `grc` element of the `spec.overrides` in the `MultiClusterHub` first and then run the loop with `"enabled":true`. If the deployments are unsuccessful, the `governance-policy-addon-controller` might have a stale lease. Delete the lease by using the following command:

Reviewer 2: hmmm, wonder do we need re-enable?

Reviewer 2: This seems like multiple steps in one step that maybe should be broken out. If we say First and Next in one step, that is usually a sign that we are merging two steps in one and not great UX.
